sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala

```diff
@@ -88,6 +88,276 @@ case class Average(child: Expression) extends AlgebraicAggregate {
+        Cast(Literal(6), resultType) - deltaN * currentM3 * Cast(Literal(4), resultType)
+    } else {
+      Cast(Literal(0), resultType)
+    }
+
+    Seq(
+      /* currentM4 = */ If(IsNull(child), currentM4, updateM4),
+      /* currentM3 = */ If(IsNull(child), currentM3, updateM3),
+      /* currentM2 = */ If(IsNull(child), currentM2, updateM2),
+      /* currentM1 = */ If(IsNull(child), currentM1, updateM1),
+      /* currentM0 = */ If(IsNull(child), currentM0, updateM0)
+    )
+  }
+
+  override val mergeExpressions = {
+
+
+    lazy val updateM0 = currentM0.left + currentM0.right
+    lazy val delta = currentM1.right - currentM1.left
+    lazy val deltaN = delta / updateM0
+
+    lazy val updateM1 = if (maxMoment >= 1) {
+      currentM1.left + delta * (currentM0.right / updateM0)
+    } else {
+      Cast(Literal(0), resultType)
+    }
+
+    lazy val updateM2 = if (maxMoment >= 2) {
+      currentM2.left + currentM2.right + delta * deltaN * currentM0.left * currentM0.right
+    } else {
+      Cast(Literal(0), resultType)
+    }
+
+    lazy val updateM3 = if (maxMoment >= 3) {
+      currentM3.left + currentM3.right + deltaN * deltaN * delta * currentM0.left *
+        currentM0.right * (currentM0.left - currentM0.right) +
+        deltaN * (currentM0.left * currentM2.right - currentM0.right * currentM2.left) *
+          Cast(Literal(3), resultType)
+    } else {
+      Cast(Literal(0), resultType)
+    }
+
+    lazy val updateM4 = if (maxMoment >= 4) {
+      currentM4.left + currentM4.right + deltaN * deltaN * deltaN * delta * currentM0.left *
+        currentM0.right * (currentM0.left * currentM0.left - currentM0.left * currentM0.right +
+        currentM0.right * currentM0.right) + deltaN * deltaN * Cast(Literal(6), resultType) *
+        (currentM0.left * currentM0.left * currentM2.right +
+          currentM0.right * currentM0.right * currentM2.left) +
+        deltaN * Cast(Literal(4), resultType) *
+          (currentM0.left * currentM3.right - currentM0.right * currentM3.left)
+    } else {
+      Cast(Literal(0), resultType)
+    }
+
+    Seq(
+      /* currentM4 = */ If(IsNull(currentM4.left), currentM4.right,
+                          If(IsNull(currentM4.right), currentM4.left, updateM4)),
+      /* currentM3 = */ If(IsNull(currentM3.left), currentM3.right,
+                          If(IsNull(currentM3.right), currentM3.left, updateM3)),
+      /* currentM2 = */ If(IsNull(currentM2.left), currentM2.right,
+                          If(IsNull(currentM2.right), currentM2.left, updateM2)),
+      /* currentM1 = */ If(IsNull(currentM1.left), currentM1.right,
+                            If(IsNull(currentM1.right), currentM1.left, updateM1)),
+      /* currentM0 = */ If(IsNull(currentM0.left), currentM0.right,
+                            If(IsNull(currentM0.right), currentM0.left, updateM0))
+    )
+  }
+}
+
+//case class Average(child: Expression) extends CentralMomentAgg(child) {
```

Comment:
Aggregate functions which depend on lower order moments can easily be implemented using the CentralMomentAgg base class. I have commented them out for now, but left them for discussion purposes.